File: backend/app/schemas/marine.py
"""
Unified Marine Cross-Domain Schemas.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class MarineObservationItem(BaseModel):
    id: str
    domain: str  # oceanography, fisheries, biodiversity
    dataset_id: Optional[str] = None
    latitude: float
    longitude: float
    depth: Optional[float] = None
    time: Optional[str] = None
    species_id: Optional[str] = None
    species_name: Optional[str] = None
    measurements: Dict[str, Any] = Field(default_factory=dict)


class MarineSummaryResponse(BaseModel):
    total_datasets: int
    oceanography_count: int
    fisheries_count: int
    biodiversity_count: int
    total_species: int


class MarineQueryRequest(BaseModel):
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    radius_km: Optional[float] = None
    bbox: Optional[List[float]] = None  # [min_lon, min_lat, max_lon, max_lat]
    depth_min: Optional[float] = None
    depth_max: Optional[float] = None
    species_id: Optional[str] = None
    domain: Optional[str] = None
    dataset_id: Optional[str] = None
    variable: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=50, ge=1, le=100)


class CrossDomainLocationDetailResponse(BaseModel):
    coordinates: Dict[str, float]
    region: str = "Indian Ocean"
    bathymetryDepth: Optional[float] = None
    oceanography: Dict[str, Any] = Field(default_factory=dict)
    fisheries: Dict[str, Any] = Field(default_factory=dict)
    biodiversity: Dict[str, Any] = Field(default_factory=dict)
    molecularEdna: Dict[str, Any] = Field(default_factory=dict)
    aiPrediction: Optional[Dict[str, Any]] = None
    associations_summary: Optional[Dict[str, Any]] = None

File: frontend/src/types/marine.ts
import { DatasetDomain } from './dataset';

export interface MarineObservation {
  id: string;
  datasetId: string;
  domain: DatasetDomain;
  timestamp?: string;
  latitude: number;
  longitude: number;
  depthMeters?: number | null;
  stationId?: string;
  region: 'Arabian Sea' | 'Bay of Bengal' | 'Lakshadweep' | 'Andaman & Nicobar' | 'Indian Ocean';
  
  // Cross-domain values when fused
  temperature?: number; // °C
  salinity?: number; // PSU
  dissolvedOxygen?: number; // mg/L or ml/L
  chlorophyllA?: number; // mg/m³
  ph?: number;
  turbidity?: number;
  
  speciesName?: string;
  scientificName?: string;
  taxonomicGroup?: string;
  individualCount?: number;
  
  catchWeightKg?: number;
  fishingEffortHours?: number;
  gearType?: string;
  vesselType?: string;
  
  ednaDetectionsCount?: number;
  ednaTargetMarker?: string;
  ednaConfidenceScore?: number;
  
  anomalyFlag?: boolean;
  anomalyScore?: number;
  riskLevel?: 'low' | 'moderate' | 'high' | 'critical';
}

export interface MarineSpatialQuery {
  dateFrom?: string;
  dateTo?: string;
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
  bbox?: [number, number, number, number]; // [minLat, minLon, maxLat, maxLon]
  depthMin?: number;
  depthMax?: number;
  species?: string;
  domain?: DatasetDomain | 'all';
  datasetId?: string;
  variable?: string;
}

export interface MarineSummary {
  totalObservations: number;
  totalDatasets: number;
  totalSpeciesRecorded: number;
  totalEdnaDetections: number;
  activeAnomalies: number;
  spatialCoveragePercentage: number | null;
  regionsBreakdown: {
    region: string;
    observationCount: number;
    speciesCount: number;
    activeVessels: number;
  }[];
  temporalSpan: {
    earliest: string;
    latest: string;
  } | null;
}

export interface CrossDomainLocationDetail {
  coordinates: {
    latitude: number;
    longitude: number;
  };
  region: string;
  bathymetryDepth: number;
  oceanography: {
    seaSurfaceTemperature: number;
    salinity: number;
    dissolvedOxygen: number;
    chlorophyllA: number;
    thermoclineDepth: number;
    mixedLayerDepth: number;
    lastUpdated: string;
  };
  fisheries: {
    dominantCatch: string;
    totalLandingsTons: number;
    cpueKgPerHour: number;
    dominantGear: string;
    fishingPressureLevel: 'Low' | 'Moderate' | 'High' | 'Very High';
  };
  biodiversity: {
    speciesRecordedCount: number;
    keySpeciesPresent: string[];
    shannonWienerIndex: number;
    endemicSpeciesFlag: boolean;
  };
  molecularEdna: {
    samplesAnalyzed: number;
    taxaIdentified: number;
    topDetections: {
      species: string;
      confidence: number;
      marker: string;
    }[];
  };
  aiPrediction: {
    anomalyDetected: boolean;
    anomalyType?: string;
    riskScore: number;
    habitatSuitabilityPercent: number;
    recommendation: string;
  };
}
